### PR TITLE
Feature layout

### PR DESCRIPTION
**Pull request by sccolbert**
Ready for merge.

Tests disabled in the name of pragmatism for now. They will be revived before `1.0` final, but we need a new beta like *now*.

### CODE REVIEW

**Comment by sccolbert** [comment]
@afshin @blink1073 @dwillmer @jasongrout @ellisonbg

File: src/panel.ts [hunk 1]
```diff
@@ -26,23 +36,109 @@ const PANEL_CLASS = 'p-Panel';
...
+class Panel extends Widget {
+  /**
+   * Create a panel layout to use with a panel.
+   *
+   * @returns A new panel layout to use with a panel.
+   *
+   * #### Notes
+   * This may be reimplemented by a subclass to create custom layouts.
+   */
+  static createLayout(): PanelLayout {
```
Comment (blink1073) [inline src/panel.ts]: No separate namespace for this one?

File: src/panel.ts [hunk 2]
```diff
@@ -60,143 +158,149 @@ abstract class AbstractPanel extends Widget {
...
   }
 
   /**
-   * Get the number of children in the panel.
+   * Get the number of child widgets in the layout.
    *
-   * @returns The number of child widgets in the panel.
+   * @returns The number of child widgets in the layout.
    */
   childCount(): number {
```
Comment (blink1073) [inline src/panel.ts]: Why not a property here?

File: src/widget.ts [hunk 3]
```diff
@@ -388,26 +355,75 @@ class Widget extends NodeWrapper implements IDisposable, IMessageHandler {
...
     this.node.parentNode.removeChild(this.node);
   }
 
   /**
+   * Test whether the given widget flag is set.
+   *
+   * #### Notes
+   * This will not typically be consumed directly by user code.
+   */
+  testFlag(flag: WidgetFlag): boolean {
```
Comment (blink1073) [inline src/widget.ts]: Is this intended to be used by other classes?

File: src/panel.ts (hunk 1 above)
Comment (sccolbert) [inline src/panel.ts]: Nope. It's intended to be reimplemented. I'm moving things that should be `const` or not reimplemented to a merged namespace. At the end of the day, there is no physical difference (it's all just slapped onto the function), but the difference is intent of use.

File: src/panel.ts (hunk 2 above)
Comment (sccolbert) [inline src/panel.ts]: Because it's `abstract` in the base class, and only methods can be `abstract`.

File: src/widget.ts (hunk 3 above)
Comment (sccolbert) [inline src/widget.ts]: It's possible. Originally, I was thinking of adding a flag `DisallowLayoutChange`, which would have been used by external code, but now that's the only behavior for layouts, so there's no need for that flag. However, if someone were to reimplement `processMessage` in order to delay attachment, they'd need a way to set the flags.

**Comment by blink1073** [comment]
Great, :+1:

**Pull request merged**